Add support to configure adding jackson @JsonPOJOBuilder annotations to generated builders

maff opened the pull request:
Adds a new option structure to instruct record-builder to add a Jackson `@JsonPOJOBuilder` annotation to the generated builder. While this implements library-specifics, it does not add Jackson as dependency (only on the test module).

**Example usage:**

Auto-detect Jackson version (default)
```java
@RecordBuilder.Options(jackson = @RecordBuilder.JacksonConfig(jsonPOJOBuilder = true))
```

Explicitly add Jackson 2.x annotation (fail if J2 is not available):
```java
@RecordBuilder.Options(
    jackson = @RecordBuilder.JacksonConfig(
        jsonPOJOBuilder = true,
        version = JacksonVersion.JACKSON_2
    )
)
```

With custom setter prefix:
```java
@RecordBuilder.Options(
    jackson = @RecordBuilder.JacksonConfig(jsonPOJOBuilder = true),
    setterPrefix = "set"
)
```

Closes #229

maff committed:
Add support to configure adding jackson @JsonPOJOBuilder annotations to generated builders

maff commented:
@Randgalt any feedback on this approach? I would be great to have this as an option for projects relying on Jackson and still wanting to benefit from RecordBuilder's initialization & defaults logic.

semyon-levin-workato commented on record-builder-processor/src/main/java/io/soabase/recordbuilder/processor/InternalRecordBuilderProcessor.java:
```diff
@@ -196,6 +198,18 @@ private void addVisibility(boolean builderIsInRecordPackage, Set<Modifier> modif
+            return;
+        }
+
+        final var annotationSpec = AnnotationSpec
```
I'm afraid it's not that simple. Recently, Jackson 3 was released with changed package names. Is it possible to check what annotations are presented on the classpath and add appropriate annotation here?

maff committed:
Update Jackson support to support both Jackson 2 and 3

Introduces a dedicated Jackson option structure. JSONPojoBuilder support
can be enabled with a boolean flag.

Detects available jackson version on the classpath and adds the
respective annotations when enabled. Jackson version can be set to AUTO
(adds annotation for every found version) or to JACKSON_2 or JACKSON_3
(only adds annotations for defined version, fails the build when library
is not on the classpath).

maff commented:
@Randgalt @semyon-levin-workato Updated to support both Jackson 2 and 3 by detecting available classes on the classpath, with support to specify a specific version in case both versions are present.

Would be awesome if we could progress with this feature for better out-of-the-box Jackson compatibility.

semyon-levin-workato commented:
@Randgalt could you please look into this?

Randgalt commented on record-builder-processor/src/main/java/io/soabase/recordbuilder/processor/JacksonSupport.java:
```diff
@@ -0,0 +1,112 @@
+
+        switch (metaData.jackson().version()) {
+        case AUTO -> {
+            if (!jackson2Present && !jackson3Present) {
+                processingEnv.getMessager().printMessage(ERROR,
+                        "jackson.jsonPOJOBuilder is enabled but Jackson is not found on classpath. "
+                                + "Add jackson-databind dependency or disable jsonPOJOBuilder.");
+                return;
+            }
+
+            if (jackson2Present) {
+                addJacksonAnnotations(metaData, builder, JACKSON_2_ANNOTATION_PACKAGE);
+            }
+
+            if (jackson3Present) {
+                addJacksonAnnotations(metaData, builder, JACKSON_3_ANNOTATION_PACKAGE);
+            }
+        }
+
+        case JACKSON_2 -> {
+            if (!jackson2Present) {
+                processingEnv.getMessager().printMessage(ERROR,
+                        "jackson.version is set to JACKSON_2 but Jackson 2.x is not found on classpath. "
+                                + "Add jackson-databind 2.x dependency or change version to AUTO.");
+                return;
+            }
+
+            addJacksonAnnotations(metaData, builder, JACKSON_2_ANNOTATION_PACKAGE);
+        }
+
+        case JACKSON_3 -> {
+            if (!jackson3Present) {
+                processingEnv.getMessager().printMessage(ERROR,
+                        "jackson.version is set to JACKSON_3 but Jackson 3.x is not found on classpath. "
+                                + "Add jackson-databind 3.x dependency or change version to AUTO.");
+                return;
+            }
+
+            addJacksonAnnotations(metaData, builder, JACKSON_3_ANNOTATION_PACKAGE);
+        }
```
Let's remove those returns

```suggestion
        case AUTO -> {
            if (!jackson2Present && !jackson3Present) {
                processingEnv.getMessager().printMessage(ERROR,
                        "jackson.jsonPOJOBuilder is enabled but Jackson is not found on classpath. "
                                + "Add jackson-databind dependency or disable jsonPOJOBuilder.");
            }
            else if (jackson2Present) {
                addJacksonAnnotations(metaData, builder, JACKSON_2_ANNOTATION_PACKAGE);
            }
            else /*if (jackson3Present)*/ {
                addJacksonAnnotations(metaData, builder, JACKSON_3_ANNOTATION_PACKAGE);
            }
        }

        case JACKSON_2 -> {
            if (!jackson2Present) {
                processingEnv.getMessager().printMessage(ERROR,
                        "jackson.version is set to JACKSON_2 but Jackson 2.x is not found on classpath. "
                                + "Add jackson-databind 2.x dependency or change version to AUTO.");
            }
            else {
                addJacksonAnnotations(metaData, builder, JACKSON_2_ANNOTATION_PACKAGE);
            }
        }

        case JACKSON_3 -> {
            if (!jackson3Present) {
                processingEnv.getMessager().printMessage(ERROR,
                        "jackson.version is set to JACKSON_3 but Jackson 3.x is not found on classpath. "
                                + "Add jackson-databind 3.x dependency or change version to AUTO.");
            } else {
                addJacksonAnnotations(metaData, builder, JACKSON_3_ANNOTATION_PACKAGE);
            }
        }
```

Randgalt commented on pom.xml:
```diff
@@ -171,7 +172,12 @@
-                <version>${jackson-version}</version>
+                <version>${jackson2-version}</version>
+            </dependency>
+            <dependency>
```
nit: add a new line here

Randgalt commented on record-builder-core/src/main/java/io/soabase/recordbuilder/core/RecordBuilder.java:
Please squash the two commits

Randgalt commented on record-builder-core/src/main/java/io/soabase/recordbuilder/core/RecordBuilder.java:
```diff
@@ -378,6 +403,29 @@ enum ConcreteSettersForOptionalMode {
+     * Specifies which Jackson version(s) to use when generating builder annotations.
+     */
+    enum JacksonVersion {
+        /**
```
Let's add a `NONE` attribute here and remove `@JacksonConfig`. I'd also rename this `RecordBuilderJacksonVersion` or something to avoid name collisions with other projects (I realize some other names in here suffer from this).

Randgalt commented on record-builder-core/src/main/java/io/soabase/recordbuilder/core/RecordBuilder.java:
```diff
@@ -359,6 +359,31 @@
+     */
+    @Retention(RetentionPolicy.CLASS)
+    @Target(ElementType.ANNOTATION_TYPE)
+    @interface JacksonConfig {
```
See comment below. We can remove this annotation.

Randgalt commented on record-builder-core/src/main/java/io/soabase/recordbuilder/core/RecordBuilder.java:
```diff
@@ -359,6 +359,31 @@
+         *
+         * @see JacksonConfig
+         */
+        JacksonConfig jackson() default @JacksonConfig;
```
```suggestion
JacksonVersion jackson() default JacksonVersion.NONE;
```

Randgalt commented on record-builder-processor/src/main/java/io/soabase/recordbuilder/processor/JacksonSupport.java:
```diff
@@ -0,0 +1,112 @@
+        jackson3Present = isAnnotationClassPresent(JACKSON_3_ANNOTATION_PACKAGE, JSON_POJO_BUILDER);
+    }
+
+    private boolean isAnnotationClassPresent(String packageName, String className) {
```
`className` is always `JSON_POJO_BUILDER`

Randgalt commented on record-builder-processor/src/main/java/io/soabase/recordbuilder/processor/JacksonSupport.java:
```diff
@@ -0,0 +1,112 @@
+    private final boolean jackson2Present;
+    private final boolean jackson3Present;
+
+    JacksonSupport(ProcessingEnvironment processingEnv) {
```
Given how this is used it would be better as a utility class with a private ctor and the methods becoming `static`.

Randgalt commented on record-builder-processor/src/main/java/io/soabase/recordbuilder/processor/JacksonSupport.java:
```diff
@@ -0,0 +1,112 @@
+        return processingEnv.getElementUtils().getTypeElement(packageName + "." + className) != null;
+    }
+
+    public void addJacksonAnnotations(RecordBuilder.Options metaData, TypeSpec.Builder builder) {
```
can be package-private

Randgalt commented on record-builder-processor/src/main/java/io/soabase/recordbuilder/processor/JacksonSupport.java:
```diff
@@ -0,0 +1,112 @@
+
+    public void addJacksonAnnotations(RecordBuilder.Options metaData, TypeSpec.Builder builder) {
+        // return without further processing if no annotation is enabled
+        if (!anyJacksonAnnotationEnabled(metaData)) {
+            return;
+        }
```
This is unnecessary right?

Randgalt commented on record-builder-processor/src/main/java/io/soabase/recordbuilder/processor/JacksonSupport.java:
```diff
@@ -0,0 +1,112 @@
+
+    private void addJsonPOJOBuilderAnnotation(RecordBuilder.Options metaData, TypeSpec.Builder builder,
+            String packageName) {
+        final var annotationSpec = AnnotationSpec.builder(ClassName.get(packageName, JSON_POJO_BUILDER))
```
We also need to consider `metaData.builderMethodName()` and `metaData.withClassMethodPrefix()` right? `JsonPOJOBuilder` has attributes for that.

Randgalt commented:
My apologies for how long this has taken but I just haven't had a lot of time recently. I've done a first pass review. In general I'm very positive on this. Thank you.
